### ISSUE TITLE
[Duplicate #233] Singletons not shared with factory

### ISSUE THREAD

**Issue by oliverdaniell** [boost-ext/di#272]
The example code below should create a single instance of my_singleton. The calls to factory.create() share one instance of my_singleton. However a different instance of my_singleton is passed to app.

```c++
struct my_singleton {
  my_singleton() {}
};

void assert_if_multiple_instances(my_singleton &s) {
  static my_singleton *address_of_my_singleton = &s;
  assert(address_of_my_singleton == &s);
}

struct interface1 {
  interface1(my_singleton &s, int i) { assert_if_multiple_instances(s); }
};

struct app {
  app(ifactory<interface1, int> &factory, my_singleton &s) {
    assert_if_multiple_instances(s);
    factory.create(1);
    factory.create(2);
  }

  void run() {}
};

int main() {
  namespace di = boost::di;
  auto injector = boost::di::make_injector(
      di::bind<ifactory<interface1, int>>.to(factory<interface1>{}));

  injector.create<app>().run();
}
```

Apologies if this is a duplicate of #233.

**Comment by kanstantsin-chernik** [comment]
@oliverdaniell indeed  it is duplicate. You can fix it by using `extensible_injector` extension or just wait a bit. As soon as we split extensions to .hpp and .cpp I am going to do it myself.

**Comment by oliverdaniell** [comment]
@kanstantsin-chernik I've changed the factory implementation to use extensible injectors but this does not fix the bug. Looking at the extensible injector example, the singletons are only shared because of

    di::bind<implementation3>().to(im3_orig)

If you remove that line then this assertion fails

     assert(extended_example.im3_ == orig_example.im3_);

Going back to my example, every object created by the factory shares the same instance of my_singleton, but the app class gets a different instance of my_singleton.

**Comment by Shelim** [comment]
There is still a problem I reported as #240 -> factory uses static member for shared_ptr instances. I've coded ugly workaround (see last comment of #240 ) but it is no longer valid with `extensible_injector` anymore. Any way to have this sorted out? :-)

**Comment by kanstantsin-chernik** [comment]
Fixed with `make_extensible` at #283
